### ISSUE TITLE
Generic Unbound TypeVar becoming Any

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3194]
**Describe the bug**
self.message in the code below is inferred as Any when I would expect it to be inferred as T. self.message_type is correctly inferred as type[T].

```python
from __future__ import annotations

from abc import abstractmethod
from typing import Generic, TypeVar

from typing_extensions import reveal_type

T = TypeVar('T')


class Foo(Generic[T]):
    def __init__(self) -> None:
        super().__init__()
        message_t = self.message_type() # Good message_type is type[T]
        self.message = message_t()
        reveal_type(self.message) # For some reason self.message is Any instead of T.

    @abstractmethod
    def message_type(self) -> type[T]:
        ...
```

**Expected behavior**
self.message being inferred as T. The difference becomes more meaningful when you also have code like,

```python
class SubFoo(Foo[int]):
  ...

```
where now self.message should be int, but it's still Any.

**VS Code extension or command-line**
pyright 1.1.229 and also pylance latest. I discovered this from pylance and being surprised by missing autocompletes in subclasses of Foo that specified T.

Also title is because when I add bound to the type var the issue goes away.

**Comment by erictraut** [comment]
When there is no type bound to the type, `Type[T]` is effectively just an instance of `type`, and it has a `__call__` method that is annotated to return an `Any`. That's where the `Any` return type comes from.

I've added special-case logic for this path. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.230, which I just published. It will also be included in the next release of pylance.
